src/Simplex/Messaging/Crypto.hs

```diff
@@ -139,29 +165,23 @@ aesKeySize = 256 `div` 8
 authTagSize :: Int
 authTagSize = 128 `div` 8
 
-generateKeyPair :: Int -> IO KeyPair
+generateKeyPair :: PrivateKey k => Int -> IO (KeyPair k)
 generateKeyPair size = loop
   where
     publicExponent = findPrimeFrom . (+ 3) <$> generateMax pubExpRange
-    privateKey s n d = PrivateKey {private_size = s, private_n = n, private_d = d}
     loop = do
-      (pub, priv) <- R.generate size =<< publicExponent
-      let s = R.public_size pub
-          n = R.public_n pub
-          d = R.private_d priv
-       in if d * d < n
-            then loop
-            else return (PublicKey pub, privateKey s n d)
-
-validKeyPair :: KeyPair -> Bool
-validKeyPair
-  ( PublicKey R.PublicKey {public_size, public_n = n, public_e = e},
-    PrivateKey {private_size, private_n, private_d = d}
-    ) =
-    let m = 30577
-     in public_size == private_size
-          && n == private_n
-          && m == expFast (expFast m d n) e n
+      (k, pk) <- R.generate size =<< publicExponent
+      let n = R.public_n k
+          d = R.private_d pk
+      if d * d < n
+        then loop
+        else pure (PublicKey k, mkPrivateKey pk)
+
+rsaPrivateSize :: PrivateKey k => k -> Int
+rsaPrivateSize = R.public_size . R.private_pub . rsaPrivateKey
+
+validKeyPair :: FullKeyPair -> Bool
+validKeyPair (PublicKey k, FullPrivateKey pk) = k == R.private_pub pk
```

Comment:
Maybe just remove this and only store private key? It's FullPrivateKey stored now anyway. Or do we need pubkey saved separately for re-retrieval if it was lost by the server owner? I believe we discussed it but I don't remember the reason.